presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/TemporalFunctionFactory.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.storage.organization;
+
+import com.facebook.presto.raptor.storage.StorageManagerConfig;
+import com.facebook.presto.spi.ConnectorSession;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.type.SqlDate;
+import com.facebook.presto.spi.type.SqlTimestamp;
+import com.facebook.presto.spi.type.Type;
+import org.joda.time.DateTimeZone;
+
+import javax.inject.Inject;
+
+import java.time.Duration;
+
+import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.facebook.presto.spi.type.TimestampType.TIMESTAMP;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class TemporalFunctionFactory
```

Comment:
In fact, in the simplest approach you don't need a factory. The logic is stateless, so you can pass the type as part of method call.

Reply:
Passing type as part of method call might mean we need to do type dispatching each time we call the getDay function, which might add unnecessary cost because we need to do this for each row of the page.

Reply:
I'm not concerned about an `if` check here, as that's insignificant compared to all the other work being done.

@AlekseiS, I'm interested in your idea. Can you explain more? Where would the time zone config go (which is now in the factory)?

Reply:
The comment above was about making the class directly injectable. It should be possible given that it can be made essentially stateless (except for timezone config which is constant). In this case you don't need a factory and can directly use the "timezone shard splitter" class directly. The class will have only two simple methods and will be helpful to consolidate shard boundary timezone conversion logic.